src/eventbridge/eventpattern/index.ts

```diff
@@ -0,0 +1,798 @@
+            : isEmptyPattern(pattern1)
+            ? []
+            : [pattern1]),
+          ...(isAggregatePattern(pattern2)
+            ? pattern2.patterns
+            : isEmptyPattern(pattern2)
+            ? []
+            : [pattern2]),
+        ],
+      };
+    }
+    // we don't know how to do this (yet?)
+    throw new Error(`Patterns of key ${key} defined at different levels.`);
+  };
+
+  const handleNumericRange = (expr: BinaryExpr): PatternDocument => {
+    const { eventReference, eventExpr, other, op } = assertOneEventReference(
+      expr.left,
+      expr.right,
+      expr.op
+    );
+
+    if (eventExpr.type !== "number") {
+      throw new Error(
+        "Numeric range only supported for event properties of type number"
+      );
+    }
+
+    const value = assertNumber(getConstant(other)?.value);
+
+    assertValidEventRefererence(eventReference);
+    const range = createSingleNumericRange(value, op);
+    if (range) {
+      return eventReferenceToPatternDocument(eventReference, range);
+    }
+
+    throw Error(`Unsupported numeric range operation: ${op}.`);
+  };
+
+  /**
+   * Any call that we encouter.
+   * Event bridge pattern match calls must all be results at build time.
+   *
+   * Limited to just Includes and StartsWith.
+   */
+  const handleCall = (expr: CallExpr): PatternDocument => {
+    if (
+      // the expr of the call
+      isPropAccessExpr(expr.expr) ||
+      isElementAccessExpr(expr.expr)
+    ) {
+      const operation = getPropertyAccessKey(expr.expr);
+
+      if (operation === OPERATIONS.INCLUDES) {
+        return handleIncludesCall(
+          expr as CallExpr & { expr: PropAccessExpr | ElementAccessExpr }
+        );
+      } else if (operation === OPERATIONS.STARTS_WITH) {
+        return handleStartsWithCall(
+          expr as CallExpr & { expr: PropAccessExpr | ElementAccessExpr }
+        );
+      }
+
+      throw new Error(`Unsupported operation ${operation}`);
+    }
+
+    throw new Error("Operations only supported on properties of the event.");
+  };
+
+  /**
+   * Array includes method.
+   *
+   * Does not support the start position parameter.
+   *
+   * someArray.includes(someValue)
+   */
+  const handleIncludesCall = (
+    expr: CallExpr & { expr: PropAccessExpr | ElementAccessExpr }
+  ): PatternDocument => {
+    const searchElement = getConstant(
+      expr.args[INCLUDES_SEARCH_ELEMENT]
+    )?.value;
+
+    if (
+      Object.values(expr.args).filter((e) => !isNullLiteralExpr(e)).length > 1
+    ) {
+      throw new Error("Includes only supports the searchElement argument");
+    }
+
+    // the property the call is on
+    const eventReference = getEventReference(expr.expr.expr);
+
+    if (!eventReference) {
+      throw new Error("Includes operation must be on a property of the event.");
+    }
+
+    if (
+      isPropAccessExpr(expr.expr.expr) ||
+      isElementAccessExpr(expr.expr.expr)
+    ) {
+      if (expr.expr.expr.type === "number[]") {
```

Comment:
Nice, this is pretty cool. But also makes me wish we could use the `ts.TypeChecker` and its AST in the CDK code. We'll very quickly want more type information than strings. At the very least, we will need to create Expressions for representing type information in a structured form. Maybe building on top of JSON schema as the underlying structural type system>

I wonder, what happens if someone is aliasing a `string` type?

```ts
type UserId = string
```

What will the type be when converted to a string? `string` or `UserId`?

Reply:
Seems to work?

```ts
type MyString = string;
interface MyTest extends EventBusRuleInput<{ s: MyString }> {}

test("non-string type", () => {
  ebEventTargetTestCase<MyTest>(
    reflect((event) => event.detail.s + event.detail.s),
    aws_events.RuleTargetInput.fromObject(
      `${EventField.fromPath("$.detail.s")}${EventField.fromPath("$.detail.s")}`
    )
  );
});
```